components/scream/src/physics/shoc/shoc_calc_shoc_vertflux_impl.hpp

```diff
@@ -0,0 +1,40 @@
+#ifndef SHOC_CALC_SHOC_VERTFLUX_IMPL_HPP
+#define SHOC_CALC_SHOC_VERTFLUX_IMPL_HPP
+
+#include "shoc_functions.hpp" // for ETI only but harmless for GPU
+
+namespace scream {
+namespace shoc {
+
+template<typename S, typename D>
+void Functions<S,D>
+::calc_shoc_vertflux(
+  const MemberType& team,
+  const Int& shcol, const Int& nlev,
+  const uview_2d<const Spack>& tkh_zi,
+  const uview_2d<const Spack>& dz_zi,
+  const uview_2d<const Spack>& invar,
+  const uview_2d<Spack>& vertflux)
+{
+  const Int nlev_pack = scream::pack::npack<Spack>(nlev);
```

Comment:
This is an interesting chunk of code. The shoc f90 is using (i) as the fast index, which makes it different from p3. We can either follow this pattern exactly in CXX and use right-layout views, or we have to re-write the code in CXX to use (k) as the fast index. I chose the latter since it made more sense to pack the cols than it did to pack rows.

Reply:
This in fact is _no_ different than P3; it's just that P3 inefficiently ignores the fact that `i` is the fast index. It passes `foo(i,:)` to a lot of functions, so `k` is not linear access in the function, which is slow. SHOC does the right thing, as do many of the established parameterizations in E3SM.

Just to be clear, we _definitely_ intend `k` to be the fast index in SCREAM. We need that to expose maximum ||ism over (k,i) while keeping the code clean. In principle, any fast index is fine as long as ||ism is over the full (k,i) space, but I think `i` as the fast index would make for a lot more code clutter, plus the length of the `i` dimension would have certain constraints we can't guarantee in practice for good performance.

Reply:
Thanks @ambrad , that's a very illuminating comment.